File: src/java/org/voegtle/weatherstation/server/OutgoingServlet.java
package org.voegtle.weatherstation.server;

import org.json.JSONObject;
import org.voegtle.weatherstation.server.data.RainDTO;
import org.voegtle.weatherstation.server.data.Statistics;
import org.voegtle.weatherstation.server.data.UnformattedWeatherDTO;
import org.voegtle.weatherstation.server.logic.WeatherDataFetcher;
import org.voegtle.weatherstation.server.persistence.AggregatedWeatherDataSet;
import org.voegtle.weatherstation.server.persistence.SmoothedWeatherDataSet;
import org.voegtle.weatherstation.server.request.DataType;
import org.voegtle.weatherstation.server.request.OutgoingUrlParameter;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class OutgoingServlet extends AbstractServlet {

  @Override
  public void init() throws ServletException {
    super.init();
  }

  @Override
  public void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
    response.setHeader("Access-Control-Allow-Origin", "*");

    WeatherDataFetcher weatherDataFetcher = new WeatherDataFetcher(pm, locationProperties);

    OutgoingUrlParameter param = new OutgoingUrlParameter(request, locationProperties.getDateUtil());
    boolean authorized = isReadSecretValid(param.getSecret());

    if (param.getType() == DataType.AGGREGATED) {
      List<AggregatedWeatherDataSet> result = weatherDataFetcher.getAggregatedWeatherData(param.getBegin(), param.getEnd());
      returnAggregatedResult(response, result, param.isExtended());
    } else if (param.getType() == DataType.CURRENT) {
      UnformattedWeatherDTO currentWeatherData = weatherDataFetcher.getLatestWeatherDataUnformatted(authorized);
      returnCurrentWeatherData(response, currentWeatherData, param.isExtended(), param.isNewFormat());
    } else if (param.getType() == DataType.RAIN) {
      RainDTO rainData = weatherDataFetcher.fetchRainData();
      writeResponse(response, jsonConverter.toJson(rainData));
    } else if (param.getType() == DataType.STATS) {
      Statistics stats = weatherDataFetcher.fetchStatistics();
      writeResponse(response, jsonConverter.toJson(stats, param.isNewFormat()));
    } else if (param.getBegin() != null) {
      List<SmoothedWeatherDataSet> result = weatherDataFetcher.fetchSmoothedWeatherData(param.getBegin(), param.getEnd());
      returnDetailedResult(response, result, authorized);
    }

  }

  private void returnCurrentWeatherData(HttpServletResponse response, UnformattedWeatherDTO currentWeatherData, boolean extended, boolean newFormat) {
    JSONObject json = newFormat ?  jsonConverter.toJson(currentWeatherData) : jsonConverter.toJsonLegacy(currentWeatherData, extended);
    writeResponse(response, json);
  }

  private void returnAggregatedResult(HttpServletResponse response, List<AggregatedWeatherDataSet> list, boolean extended) {
    ArrayList<JSONObject> jsonObjects = jsonConverter.toJsonAggregated(list, extended);

    writeResponse(response, jsonObjects);

  }

}

File: src/main/java/org/voegtle/weatherstation/server/OutgoingServlet.kt
package org.voegtle.weatherstation.server

import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.voegtle.weatherstation.server.data.UnformattedWeatherDTO
import org.voegtle.weatherstation.server.logic.WeatherDataFetcher
import org.voegtle.weatherstation.server.persistence.entities.AggregatedWeatherDataSet
import org.voegtle.weatherstation.server.request.DataType
import org.voegtle.weatherstation.server.request.OutgoingUrlParameter
import java.io.IOException
import java.util.*

class OutgoingServlet : AbstractServlet() {

  @Throws(ServletException::class, IOException::class)
  public override fun doGet(request: HttpServletRequest, response: HttpServletResponse) {
    response.setHeader("Access-Control-Allow-Origin", "*")

    val weatherDataFetcher = WeatherDataFetcher(pm, locationProperties!!)

    val param = OutgoingUrlParameter(request, locationProperties!!.dateUtil)
    val authorized = isReadSecretValid(param.secret)

    when {
      param.type == DataType.AGGREGATED -> {
        val result = weatherDataFetcher.getAggregatedWeatherData(param.begin!!, param.end ?: Date())
        returnAggregatedResult(response, result)
      }
      param.type == DataType.CURRENT -> {
        val currentWeatherData = weatherDataFetcher.getLatestWeatherDataUnformatted(authorized)
        returnCurrentWeatherData(response, currentWeatherData)
      }
      param.type == DataType.RAIN -> {
        val rainData = weatherDataFetcher.fetchRainData()
        writeResponse(response, jsonConverter!!.toJson(rainData))
      }
      param.type == DataType.STATS -> {
        val stats = weatherDataFetcher.fetchStatistics()
        writeResponse(response, jsonConverter!!.toJson(stats))
      }
      param.begin != null -> {
        val result = weatherDataFetcher.fetchSmoothedWeatherData(param.begin, param.end!!)
        log.info(
            "begin: ${param.begin} end: ${param.end} localTZ: ${param.isLocalTimezone} localTimezone: ${locationProperties?.dateUtil?.timezone?.displayName}")

        returnDetailedResult(response, result, authorized)
      }
    }

  }

  private fun returnCurrentWeatherData(response: HttpServletResponse, currentWeatherData: UnformattedWeatherDTO) {
    val json = jsonConverter!!.toJson(currentWeatherData)
    writeResponse(response, json)
  }

  private fun returnAggregatedResult(response: HttpServletResponse, list: List<AggregatedWeatherDataSet>) {
    val jsonObjects = jsonConverter!!.toJsonAggregated(list)
    writeResponse(response, jsonObjects)
  }

}
